lambda/translcore.ml

```diff
@@ -447,6 +459,75 @@ and transl_exp0 ~in_new_scope ~scopes e =
   | Texp_while(cond, body) ->
       Lwhile(transl_exp ~scopes cond,
              event_before ~scopes body (transl_exp ~scopes body))
+  
+  | Texp_arr_comprehension (_param, _body, _type_comp) -> 
+    (*
+    let type_comp = List.hd type_comp in
+    let param = List.hd param in
+    let pval, args, func = match type_comp with
+    | From_to (_,e2,e3, dir) -> 
+        Pintval, [transl_exp ~scopes e2; transl_exp ~scopes e3],
+        comprehension_prim ~is_array:true ~dir
+    | In (_, e2) ->  
+        Pgenval, [transl_exp ~scopes e2],
+        list_comp_in_prim ~is_array:true
+    in
+    let fn = Lfunction {kind = Curried;
+                        params= [param, pval];
+                        return = Typeopt.value_kind body.exp_env body.exp_type;
+                        attr = default_function_attribute;
+                        loc = Loc_unknown;
+                        body = transl_exp ~scopes  body} in
+    Lapply{
+      ap_loc=Loc_unknown;
+      ap_func=func;
+      ap_args= fn::args;
+      ap_tailcall=Default_tailcall;
+      ap_inlined=Default_inline;
+      ap_specialised=Default_specialise;
+    }
+    *)
+    assert false
+  | Texp_list_comprehension (params, body, type_comps) -> 
+    (*The left most statement should be at the end of the recursion.*)
+    let params = List.rev (List.concat params) in
+    let type_comps = List.rev (List.concat type_comps) in
+    let return = Typeopt.value_kind body.exp_env body.exp_type in
+    let create type_comp param bdy is_concat = 
+      let pval, args, func = match type_comp with
+      | From_to (_,e2,e3, dir) -> 
+          Pintval, [transl_exp ~scopes e2; transl_exp ~scopes e3],
+          comprehension_prim ~is_concat ~dir
+      | In (_, e2) ->  
+          Pgenval, [transl_exp ~scopes e2],
+          list_comp_in_prim ~is_concat
+      in
+      let fn = Lfunction {kind = Curried;
+                          params= [param, pval];
+                          return = return;
```

Comment:
I think using `return` here is only correct for the case where `bdy` equals `body`. I think in the other cases the functions instead return a list.

Reply:
Does that mean I should use Pgenval for the case when it is a list?

Reply:
Yeah, Pgenval would be right for a list.